src/eval.py

```diff
@@ -515,6 +520,104 @@ def predict_spec_sliding_window(spectrogram, model, chunk_size=256, jump=128):
+
+
+def forward_inference_on_batch(
+        model,  # the ML model we use to generate predictions
+        data,  # input data, of dims number of time segments, number of frequencies
+        time_idx,  # the beginning time index for this inference
+        jump,  # number of time steps between starts of frames to perform inference on (must be a clean divisor of TIME_STEPS_IN_WINDOW)
+        batchsize,  # number of frames to perform inference on at once. Be careful not to exceed VRAM limits! This is going to be highly hardware-dependent.
+        predictions,  # full ndarray of the sum of all predictions at each individual time step
+        overlap_counts,  # full ndarray of the number of predictions applied to each invididual time step
+
+        # the number of different offsets that should be used. This method will process this many batches of input.
+        # If this number is 1, no 'jumps' will actually be evaluated, just the standard start of the array (the offset of 0).
+        max_jumps):
+
+    # select the region of the data to perform inference on
+    input_batch = data[time_idx:(time_idx + batchsize * TIME_STEPS_IN_WINDOW + (max_jumps - 1) * jump), :]
+
+    # one batch of spectrogram 'frames' (each representing a full input to the model) to be processed in parallel are consecutive and non-overlapping.
+    # each iteration of this loop performs this approach with a different offset into the first frame, allowing the evaluation
+    # of overlapping frames. Overlapping frames are NOT evaluated in parallel, but non-overlapping consecutive frames can be.
+    for num_jumps_to_offset in range(0, max_jumps):
+        # Used for indexing into the input batch
+        local_begin_idx = num_jumps_to_offset * jump
+        local_end_idx = local_begin_idx + batchsize * TIME_STEPS_IN_WINDOW
+
+        # Used for indexing into the prediction arrays
+        global_begin_idx = local_begin_idx + time_idx
+        global_end_idx = local_end_idx + time_idx
+
+        reshaped_input_batch = input_batch[local_begin_idx:local_end_idx, :].reshape(batchsize, TIME_STEPS_IN_WINDOW, FREQS)
+
+        # apply per-frame normalization
+        # TODO: figure out how to apply per-frame normalization on the GPU/device so we can get rid of this annoying perf hit
+        # TODO: If we can do this, we won't have to copy the same data to the GPU/device 'max_jumps' times...
+        means = np.mean(reshaped_input_batch, axis=(1,2)).reshape(-1, 1, 1)
+        stds = np.std(reshaped_input_batch, axis=(1,2)).reshape(-1, 1, 1)
+
+        reshaped_input_batch -= means
+        reshaped_input_batch /= stds
+
+        reshaped_input_batch_var = Variable(torch.from_numpy(reshaped_input_batch).float().to(parameters.device))
+
+        raw_outputs = model(reshaped_input_batch_var)
+        outputs = raw_outputs.view(batchsize, TIME_STEPS_IN_WINDOW)
+
+        relevant_predictions = predictions[global_begin_idx:global_end_idx].reshape((batchsize, TIME_STEPS_IN_WINDOW))
```

Comment:
Wow, this is surprising to me that a copy is not made here when doing this slicing. No wonder weird things can happen when you don't copy data